numpy/core/src/multiarray/einsum_sumprod.c.src

```diff
@@ -250,244 +236,198 @@ static void
-/**begin repeat2
- * #i = 6, 5, 4, 3, 2, 1, 0#
- */
-        case @i@+1:
-            data_out[@i@] = @to@(value0 *
-                                 @from@(data1[@i@]) +
-                                 @from@(data_out[@i@]));
-/**end repeat2**/
-        case 0:
-            return;
-    }
-
-#if EINSUM_USE_SSE1 && @float32@
-    value0_sse = _mm_set_ps1(value0);
-
+#if @NPYV_CHK@ // NPYV check for @type@
     /* Use aligned instructions if possible */
-    if (EINSUM_IS_SSE_ALIGNED(data1) && EINSUM_IS_SSE_ALIGNED(data_out)) {
-        /* Unroll the loop by 8 */
-        while (count >= 8) {
-            count -= 8;
-
-/**begin repeat2
- * #i = 0, 4#
- */
-            a = _mm_mul_ps(value0_sse, _mm_load_ps(data1+@i@));
-            b = _mm_add_ps(a, _mm_load_ps(data_out+@i@));
-            _mm_store_ps(data_out+@i@, b);
-/**end repeat2**/
-            data1 += 8;
-            data_out += 8;
-        }
-
-        /* Finish off the loop */
-        if (count > 0) {
-            goto finish_after_unrolled_loop;
+    #ifndef NPY_HAVE_NEON
+    const int is_aligned = EINSUM_IS_ALIGNED(data1) && EINSUM_IS_ALIGNED(data_out);
+    #else
+    // ARM/Neon don't have instructions for aligned memory access
+    const int is_aligned = 0;
+    #endif
+    const int vstep = npyv_nlanes_@sfx@;
+    const npyv_@sfx@ va_scalar = npyv_setall_@sfx@(a_scalar);
+
+    /**begin repeat2
+     * #cond = if(is_aligned), else#
+     * #ld = loada, load#
+     * #st = storea, store#
+     */
+    @cond@ {
+    #if @unroll_by@ == 4
+        const npy_intp vstepx4 = vstep * 4;
+        for (; count >= vstepx4; count -= vstepx4, data1 += vstepx4, data_out += vstepx4) {
+            /**begin repeat3
+             * #i = 0, 1, 2, 3#
+             */
+            npyv_@sfx@ b@i@ = npyv_@ld@_@sfx@(data1 + vstep * @i@);
+            npyv_@sfx@ c@i@ = npyv_@ld@_@sfx@(data_out + vstep * @i@);
+            /**end repeat3**/
+            /**begin repeat3
+             * #i = 0, 1, 2, 3#
+             */
+            npyv_@sfx@ abc@i@ = npyv_muladd_@sfx@(va_scalar, b@i@, c@i@);
+            /**end repeat3**/
+            /**begin repeat3
+             * #i = 0, 1, 2, 3#
+             */
+            npyv_@st@_@sfx@(data_out + vstep * @i@, abc@i@);
+            /**end repeat3**/
         }
-        else {
-            return;
+    #elif @unroll_by@ == 2
+        const npy_intp vstepx2 = vstep * 2;
+        for (; count >= vstepx2; count -= vstepx2, data1 += vstepx2, data_out += vstepx2) {
+            npyv_@sfx@ b0 = npyv_@ld@_@sfx@(data1);
+            npyv_@sfx@ b1 = npyv_@ld@_@sfx@(data1 + vstep);
+            npyv_@sfx@ c0 = npyv_@ld@_@sfx@(data_out);
+            npyv_@sfx@ c1 = npyv_@ld@_@sfx@(data_out + vstep);
+            npyv_@sfx@ abc0 = npyv_muladd_@sfx@(va_scalar, b0, c0);
+            npyv_@sfx@ abc1 = npyv_muladd_@sfx@(va_scalar, b1, c1);
+            npyv_@st@_@sfx@(data_out, abc0);
+            npyv_@st@_@sfx@(data_out + vstep, abc1);
         }
+    #endif
```

Comment:
Perhaps for clarity
```suggestion
    #else
        #error "Invalid unroll_by = @unroll_by@"
    #endif
```
I know it should never happen, but it makes that obvious to the reader.

Reply:
applied.